tiktorch/server/session/process.py

```diff
@@ -1,123 +1,126 @@
+        for tensor in tensors:
+            self.check_shape(tensor.spec_id, tensor.data.dims, tensor.data.shape)
+
+    def check_shape(self, spec_id: str, axes: Tuple[str, ...], shape: Tuple[int, ...]):
+        shape = self._get_axes_with_size(axes, shape)
+        spec = self._get_input_spec(spec_id)
+        if isinstance(spec.shape, list):
+            self._check_shape_explicit(spec, shape)
+        elif isinstance(spec.shape, ParametrizedInputShape):
+            self._check_shape_parameterized(spec, shape)
+        else:
+            raise ValueError(f"Unexpected shape {spec.shape}")
+
+    def _get_input_spec(self, spec_id: str) -> nodes.InputTensor:
+        self._check_spec_exists(spec_id)
+        specs = [spec for spec in self._model.input_specs if spec.name == spec_id]
+        assert len(specs) == 1, "ids of tensor specs should be unique"
+        return specs[0]
+
+    def _check_spec_exists(self, spec_id: str):
+        spec_names = [spec.name for spec in self._model.input_specs]
+        if spec_id not in spec_names:
+            raise ValueError(f"Spec {spec_id} doesn't exist for specs {spec_names}")
+
+    def _check_shape_explicit(self, spec: nodes.InputTensor, tensor_shape: Dict[str, int]):
+        assert self._is_shape_explicit(spec)
+        reference_shape = {name: size for name, size in zip(spec.axes, spec.shape)}
+        if reference_shape != tensor_shape:
+            raise ValueError(f"Incompatible shapes found {tensor_shape}, expected {reference_shape}")
+
+    def _check_shape_parameterized(self, spec: nodes.InputTensor, tensor_shape: Dict[str, int]):
+        assert isinstance(spec.shape, ParametrizedInputShape)
+        if not self._is_shape(tensor_shape.values()):
+            raise ValueError(f"Invalid shape's sizes {tensor_shape}")
+
+        min_shape = self._get_axes_with_size(spec.axes, tuple(spec.shape.min))
+        step = self._get_axes_with_size(spec.axes, tuple(spec.shape.step))
+        assert min_shape.keys() == step.keys()
+        if tensor_shape.keys() != min_shape.keys():
+            raise ValueError(f"Incompatible axes for tensor {tensor_shape} and spec {spec}")
+
+        tensor_shapes_arr = np.array(list(tensor_shape.values()))
+        min_shape_arr = np.array(list(min_shape.values()))
+        step_arr = np.array(list(step.values()))
+        diff = tensor_shapes_arr - min_shape_arr
+        if any(size < 0 for size in diff):
+            raise ValueError(f"Tensor shape {tensor_shape} smaller than min shape {min_shape}")
+
+        non_zero_idx = np.nonzero(step_arr)
+        multipliers = diff[non_zero_idx] / step_arr[non_zero_idx]
+        multiplier = np.unique(multipliers)
+        if len(multiplier) == 1 and self._is_natural_number(multiplier[0]):
+            return
+        raise ValueError(f"Tensor shape {tensor_shape} not valid for spec {spec}")
+
+    def _is_natural_number(self, n) -> bool:
+        return np.floor(n) == np.ceil(n) and n >= 0
```

Comment:
I guess that's a bit of an edge case, but your code will return an incorrect result for `np.inf`

```suggestion
        return n % 1 == 0.0 and n >= 0
```

Reply:
Nice! I didn't think about that, thank you